### ISSUE TITLE
If gemspec specifies Ruby patch level, `bundle update` emits uninformative message

### ISSUE THREAD

**Issue by MarkDBlackwell** [bundler/bundler#4593]
I'm writing my first gem. After `bundle gem {name}`, reading RubyGems' API, I supplemented `{name}.gemspec` suitably.

`gem build {name}.gemspec` succeeded. However, `bundle update` gave this mysterious error message:

    Could not find gem '{name}' in source at `.`.
    Source contains '{name}' at: 0.1.0 ruby
Currently, RubyGems' API states 'The Ruby version can be specified to the patch level'. It documents method [Gem::Specification#required_ruby_version=](http://docs.seattlerb.org/rubygems/Gem/Specification.html#method-i-required_ruby_version-3D) as:
```
required_ruby_version=(req)
The version of Ruby required by this gem. The ruby version can be specified to the patch-level:
$ ruby -v -e 'p Gem.ruby_version'
ruby 2.0.0p247 (2013-06-27 revision 41674) [x86_64-darwin12.4.0]
#<Gem::Version "2.0.0.247">
Because patch-level is taken into account[...]
```
Following the API, I had specified my Ruby version (including patch level) in `{name}.gemspec` as follows:

    spec.required_ruby_version = '2.3.1.112'
And I'd specified my Ruby version (including patch level) in `Gemfile` as follows:

    ruby '2.3.1', engine_version: '2.3.1', engine: 'ruby', patchlevel: '112'
Ultimately, divide-and-conquer determined Bundler seemed to be confused by the four-numbered Ruby version specification (in `{name}.gemspec`). And `rm -f Gemfile.lock; bundle update` worked if I dropped the final, patch number.

It wasn't (and isn't) obvious to me why these two Ruby version specifications don't match (for Bundler).

Perhaps (from Bundler) a more informative error message would help other first-gem writers, who follow RubyGems' API in this way?

I've created a minimal example (using current versions) for which Bundler generates the above error message:

```bash
$ pwd
/home/mark/develop

$ bundle gem bundle-gem-error
Creating gem 'bundle-gem-error'...
MIT License enabled in config
Code of conduct enabled in config
      create  bundle-gem-error/Gemfile
      create  bundle-gem-error/.gitignore
      create  bundle-gem-error/lib/bundle/gem/error.rb
      create  bundle-gem-error/lib/bundle/gem/error/version.rb
      create  bundle-gem-error/bundle-gem-error.gemspec
      create  bundle-gem-error/Rakefile
      create  bundle-gem-error/README.md
      create  bundle-gem-error/bin/console
      create  bundle-gem-error/bin/setup
      create  bundle-gem-error/.travis.yml
      create  bundle-gem-error/test/test_helper.rb
      create  bundle-gem-error/test/bundle/gem/error_test.rb
      create  bundle-gem-error/LICENSE.txt
      create  bundle-gem-error/CODE_OF_CONDUCT.md
Initializing git repo in /home/mark/develop/bundle-gem-error

$ nano bundle-gem-error/Gemfile
$ nano bundle-gem-error/bundle-gem-error.gemspec

$ cd bundle-gem-error
RVM used your Gemfile for selecting Ruby, it is all fine - Heroku does that too,
you can ignore these warnings with 'rvm rvmrc warning ignore /home/mark/develop/bundle-gem-error/Gemfile'.
To ignore the warning for all files run 'rvm rvmrc warning ignore allGemfiles'.

ruby-2.3.1,engine_version:2.3.1,engine:ruby-p112 is not installed.
To install do: 'rvm install ruby-2.3.1,engine_version:2.3.1,engine:ruby-p112'

$ pwd
/home/mark/develop/bundle-gem-error

$ ls
bin  bundle-gem-error.gemspec  CODE_OF_CONDUCT.md  Gemfile  lib  LICENSE.txt  Rakefile  README.md  test

$ gem build bundle-gem-error.gemspec
  Successfully built RubyGem
  Name: bundle-gem-error
  Version: 0.1.0
  File: bundle-gem-error-0.1.0.gem

$ gem env
RubyGems Environment:
  - RUBYGEMS VERSION: 2.6.4
  - RUBY VERSION: 2.3.1 (2016-04-26 patchlevel 112) [x86_64-linux]
  - INSTALLATION DIRECTORY: /home/mark/.rvm/gems/ruby-2.3.1
  - USER INSTALLATION DIRECTORY: /home/mark/.gem/ruby/2.3.0
  - RUBY EXECUTABLE: /home/mark/.rvm/rubies/ruby-2.3.1/bin/ruby
  - EXECUTABLE DIRECTORY: /home/mark/.rvm/gems/ruby-2.3.1/bin
  - SPEC CACHE DIRECTORY: /home/mark/.gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /home/mark/.rvm/rubies/ruby-2.3.1/etc
  - RUBYGEMS PLATFORMS:
    - ruby
    - x86_64-linux
  - GEM PATHS:
     - /home/mark/.rvm/gems/ruby-2.3.1
     - /home/mark/.rvm/gems/ruby-2.3.1@global
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /home/mark/.rvm/gems/ruby-2.3.1/bin
     - /home/mark/.rvm/gems/ruby-2.3.1@global/bin
     - /home/mark/.rvm/rubies/ruby-2.3.1/bin
     - /usr/local/bin
     - /usr/bin
     - /bin
     - /usr/local/games
     - /usr/games
     - /home/mark/.rvm/bin

$ rvm --version
rvm 1.27.0 (latest) by Wayne E. Seguin <wayneeseguin@gmail.com>, Michal Papis <mpapis@gmail.com> [https://rvm.io/]

$ cat /etc/debian_version
8.4

$ lsb_release -da
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 8.4 (jessie)
Release:	8.4
Codename:	jessie

$ hostnamectl
   Static hostname: mycroftdebian
         Icon name: computer-vm
           Chassis: vm
        Machine ID: 942aa37ac16443c5b7f04bc7491aa97c
           Boot ID: d97dbd7eb01a42b1b6d5732f72228f3c
    Virtualization: kvm
  Operating System: Debian GNU/Linux 8 (jessie)
            Kernel: Linux 3.16.0-4-amd64
      Architecture: x86-64

$ cat /proc/version
Linux version 3.16.0-4-amd64 (debian-kernel@lists.debian.org) (gcc version 4.8.4 (Debian 4.8.4-1) ) #1 SMP Debian 3.16.7-ckt25-2 (2016-04-08)
```

**Comment by segiddins** [comment]
Pretty sure this is an issue with the Resolver using a mix of `Gem::Version` and `Bundler::RubyVersion`, which would be fixed by using only one or the other

**Comment by MarkDBlackwell** [comment]
Here's more information:
```
$ rm -f Gemfile.lock; bundle update
Fetching gem metadata from https://rubygems.org/
Fetching version metadata from https://rubygems.org/
Could not find gem 'bundle-gem-error' in source at `.`.
Source contains 'bundle-gem-error' at: 0.1.0 ruby

$ bundle platform
Your platform is: x86_64-linux

Your app has gems that work on these platforms:
* ruby

Your Gemfile specifies a Ruby version requirement:
* ruby 2.3.1p112

Your current platform satisfies the Ruby version requirement.

$ bundle env
Environment

    Bundler   1.12.4
    Rubygems  2.6.4
    Ruby      2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]
    GEM_HOME  /home/mark/.rvm/gems/ruby-2.3.1
    GEM_PATH  /home/mark/.rvm/gems/ruby-2.3.1:/home/mark/.rvm/gems/ruby-2.3.1@global
    RVM       1.27.0 (latest)
    Git       2.1.4
    rubygems-bundler (1.4.4)

Bundler settings

    gem.test
      Set for the current user (/home/mark/.bundle/config): "minitest"
    gem.mit
      Set for the current user (/home/mark/.bundle/config): "true"
    gem.coc
      Set for the current user (/home/mark/.bundle/config): "true"
    orig_path
      Set via BUNDLE_ORIG_PATH: "/home/mark/.rvm/gems/ruby-2.3.1/bin:/home/mark/.rvm/gems/ruby-2.3.1@global/bin:/home/mark/.rvm/rubies/ruby-2.3.1/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games:/home/mark/.rvm/bin"
    orig_gem_path
      Set via BUNDLE_ORIG_GEM_PATH: "/home/mark/.rvm/gems/ruby-2.3.1:/home/mark/.rvm/gems/ruby-2.3.1@global"

Gemfile

    source 'https://rubygems.org'
    # ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]
    ruby '2.3.1', engine_version: '2.3.1', engine: 'ruby', patchlevel: '112'
    # Specify your gem's dependencies in bundle-gem-error.gemspec
    gemspec

Gemfile.lock

    <No /home/mark/develop/bundle-gem-error/Gemfile.lock found>

bundle-gem-error.gemspec

    # coding: utf-8
    lib = File.expand_path('../lib', __FILE__)
    $LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
    require 'bundle/gem/error/version'
    Gem::Specification.new do |spec|
    # ruby 2.3.1p112 (2016-04-26 revision 54768) [x86_64-linux]
      spec.required_ruby_version = '2.3.1.112'
      spec.name          = "bundle-gem-error"
      spec.version       = Bundle::Gem::Error::VERSION
      spec.authors       = ["fake"]
      spec.email         = ["fake@example.com"]
      spec.summary       = %q{fake}
      spec.description   = %q{fake2}
      spec.homepage      = "http://fake.example.com"
      spec.license       = "MIT"
      # Prevent pushing this gem to RubyGems.org. To allow pushes either set the 'allowed_push_host'
      # to allow pushing to a single host or delete this section to allow pushing to any host.
      if spec.respond_to?(:metadata)
        spec.metadata['allowed_push_host'] = "TODO: Set to 'http://mygemserver.com'"
      else
        raise "RubyGems 2.0 or newer is required to protect against public gem pushes."
      end
      spec.files         = `git ls-files -z`.split("\x0").reject { |f| f.match(%r{^(test|spec|features)/}) }
      spec.bindir        = "exe"
      spec.executables   = spec.files.grep(%r{^exe/}) { |f| File.basename(f) }
      spec.require_paths = ["lib"]
      spec.add_development_dependency "bundler", "~> 1.12"
      spec.add_development_dependency "rake", "~> 10.0"
      spec.add_development_dependency "minitest", "~> 5.0"
    end
```

**Comment by segiddins** [comment]
Thanks @MarkDBlackwell :) that should be enough info for us to come up with a failing test case!

**Comment by segiddins** [comment]
@chalkos any interest in trying to reduce this to a test case and then fixing it?

**Comment by chalkos** [comment]
@segiddins ok, I can look into it. Will this be a breaking change for 2.0 or a change for the next bundler 1.x version?

**Comment by segiddins** [comment]
Probably a bug fix for 1.13
